### DISCUSSION TITLE
mdx guide improvements

### DISCUSSION THREAD

**Discussion by jdevries3133** [remix-run/remix#4368] [Proposals]
I'd like to suggest some improvements to the mds guide in documentation. Consider this section:

https://github.com/remix-run/remix/blob/737f5182c0086d761f618535ceb2f344398e4766/docs/guides/mdx.md?plain=1#L191-L201

Why not recommend this much more succinct pattern instead

```javascript
exports.mdx = async (filename) => ({
  rehypePlugins: [(await import('rehype-highlight')],
  remarkPlugins: [(await import('remark-toc')]
})
```

Also, this guide doesn't specify how to pass options to plugins, as the mdx guide does: https://mdxjs.com/docs/extending-mdx/#using-plugins

Does remix use the same pattern? Or, I need to wrap the plugins like:

```javascript
exports.mdx = async () => ({
  rehypePlugins: [
    async () => {
      const hi = await import('rehype-highlight')
      return hi.default({
        subset: ['javascript', 'yaml']
      })
    }
  ]
})
```

This is something I'm trying to figure out right now, and I find myself in the remix source code to try to find the answer. I'm sure I'll figure it out, but I think this should be included on the mdx guide documentation page as well.

I'd be happy to make a contribution if others share my point of view. I think that at least documenting how to pass configuration to these plugins is the first priority. Whether or not the expression-based syntax (`[(await import(...))]`) is preferable is up for debate, so I'd probably aim for an MR achieving the first priority first.

# Update

I figured out how to configure plugins :)

I still think this is the pattern we should advise in our docs:

```javascript
  mdx: async () => ({
    rehypePlugins: [
      [(await import("rehype-highlight")).default, {
        languages: {
          vim: (await import('highlight.js/lib/languages/vim')).default
        }
      }],
      (await import("rehype-slug")).default,
      (await import("rehype-autolink-headings")).default,
      (await import("rehype-toc")).default,
    ],
  }),
```

**Comment by benseitz** [comment]
I searched how to add option to rehype and remark plugins and found this. So first of all: thank you @jdevries3133 for figuring it out and documenting it here.
Secondly: I agree that at least that should be part of the documentation. Most plugins are useless without being configured and the documentation in the plugin's READMEs isn't helpful at all.

**Reply by jdevries3133** [reply]
I'd be happy to write a PR if a maintainer is available to get it landed!

**Reply by 0xDAEF0F** [reply]
Hi @jdevries3133,

Are you sure this is the way?
I have been cracking my head for hours and I can't seem to make it work. Even tried the vim option and didn't work either.

[ERROR] Unknown language: `haskell` is not registered [plugin remix-mdx]

some-route.mdx
\`\`\`haskell
Some haskell code
\`\`\`

remix.config.js
```javascript
/** @type {import('@remix-run/dev').AppConfig} */
module.exports = {
  ignoredRouteFiles: ['**/.*'],
  mdx: async () => ({
    rehypePlugins: [
      (await import("rehype-highlight")).default, {
        languages: {
          haskell: (await import("highlight.js/lib/languages/haskell")).default
        }
      }
    ]
  }),
  tailwind: true,
};
